core/threshold/src/networking/tls.rs

```diff
@@ -36,63 +40,173 @@ pub struct ReleasePCRValues {
+    }
 }
 
 /// Our custom verifier for our custom mTLS certificates extended with AWS Nitro
 /// attestation documents. It doesn't reimplement normal X.509 certificate
-/// verification and wraps around a well-tested verifier. In addition to the usual X.509 checks, it checks PCR values from the bundled attestation document.
+/// verification and wraps around the well-tested
+/// WebPki[Client|Server]Verifier. In addition to the usual X.509 checks, it
+/// checks PCR values from the bundled attestation document. It also supports
+/// multiple trust root sets configurable at runtime which is handy when working
+/// with multiple MPC contexts.
+///
+/// The TLS certificates are expected to embed the context id in their serial
+/// number. Depending on the context id and the certificate subject name, this
+/// verifier will choose a verifier with just one appropriate CA certificate in
+/// the trust root store to actually verify the certificate.
 #[derive(Debug)]
-pub struct AttestedServerVerifier {
-    verifier: Arc<dyn ServerCertVerifier>,
-    release_pcrs: Arc<Vec<ReleasePCRValues>>,
+pub struct AttestedVerifier {
+    root_hint_subjects: Vec<DistinguishedName>,
+    supported_algs: WebPkiSupportedAlgorithms,
+    // SessionId is supposed to be based on RequestId, and we're representing
+    // ContextId as RequestId so far, so let's say it's all the same for now
+    contexts: RwLock<HashMap<SessionId, Arc<AttestedTLSContext>>>,
+    // are enclave images signed?
     pcr8_expected: bool,
 }
 
-impl AttestedServerVerifier {
-    pub fn new(
-        verifier: Arc<dyn ServerCertVerifier>,
-        release_pcrs: Arc<Vec<ReleasePCRValues>>,
-        pcr8_expected: bool,
-    ) -> Self {
-        Self {
-            verifier,
-            release_pcrs,
+impl AttestedVerifier {
+    pub fn new(pcr8_expected: bool) -> anyhow::Result<Self> {
+        Ok(Self {
+            root_hint_subjects: Vec::new(),
+            supported_algs: CryptoProvider::get_default()
+                .ok_or(anyhow!(
+                    "
+Crypto provider should exist at this point"
+                ))?
+                .signature_verification_algorithms,
+            contexts: RwLock::new(HashMap::new()),
             pcr8_expected,
-        }
+        })
+    }
+
+    pub fn add_context(
+        &self,
+        context_id: SessionId,
+        ca_certs: HashMap<String, Pem>,
+        release_pcrs: Option<Arc<Vec<ReleasePCRValues>>>,
+    ) -> anyhow::Result<()> {
+        let mut contexts = self.contexts.write().unwrap();
+        ensure!(
+            !contexts.contains_key(&context_id),
+            "Context with ID {context_id} already exists"
+        );
+        let context = AttestedTLSContext::new(ca_certs, release_pcrs)
+            .map_err(|e| anyhow::anyhow!("Failed to add new TLS context: {e}"))?;
+        contexts.insert(context_id, Arc::new(context));
+        Ok(())
+    }
+
+    pub fn remove_context(&self, context_id: SessionId) -> anyhow::Result<()> {
```

Comment:
Why is the return type a `Result` when there is no way an error could be produced in the method? Also I am wondering if it is safe to use the `std::mutex` here and panic if it is already locked. What I mean is that can we be sure that the `AttestedVerifier` is not not shared between multiple concurrent threads without a mutex? (Since locking does not require the object `AttestedVerifier` to be `mut` it could be shared between threads with an `Arc` and hence, in theory, multiple threads could be modifying it concurrently (unless I am missing something)).

Reply:
changed it here 9cb9508ddab961d453793fa0baf20d9923674c99 to return error instead of unwrap